src/main/java/org/zalando/nakadi/service/ConsumerLimitingService.java

```diff
@@ -0,0 +1,175 @@
+            // in a case of failure release slots for partitions that already acquired slots
+            slots.stream()
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .forEach(this::releaseConnectionSlot);
+            if (e instanceof NoConnectionSlotsException) {
+                throw e;
+            } else {
+                throw new ServiceUnavailableException("Error communicating with zookeeper", e);
+            }
+        }
+    }
+
+    public void releaseConnectionSlots(final List<ConnectionSlot> connectionSlots) {
+        connectionSlots.forEach(this::releaseConnectionSlot);
+    }
+
+    private void releaseConnectionSlot(final ConnectionSlot slot) {
+        final String parent = zkPathForConsumer(slot.getClient(), slot.getEventType(), slot.getPartition());
+        final String zkPath = format("{0}/{1}", parent, slot.getConnectionId());
+        try {
+            zkHolder.get()
+                    .delete()
+                    .guaranteed()
+                    .forPath(zkPath);
+            deletePartitionNodeIfPossible(parent);
+        } catch (final Exception e) {
+            LOG.error("Zookeeper error when deleting consumer connection node", e);
+        }
+    }
+
+    public void deletePartitionNodeIfPossible(final String nodePath) {
+        try {
+            zkHolder.get()
+                    .delete()
+                    .forPath(nodePath);
+        } catch (final KeeperException.NotEmptyException e) {
+            // if the node has children - we should not delete it
+        } catch (final Exception e) {
+            LOG.error("Zookeeper error when trying delete consumer node", e);
+        }
+    }
+
+    private Optional<ConnectionSlot> acquireConnectionSlot(final String client, final String eventType,
+                                                           final String partition)
+            throws ConnectionSlotOccupiedException {
+
+        final String parent = zkPathForConsumer(client, eventType, partition);
+        List<String> children = ImmutableList.of();
+        try {
+            children = zkHolder.get()
+                    .getChildren()
+                    .forPath(parent);
+        } catch (final KeeperException.NoNodeException e) {
+            // no node = no children
+        } catch (Exception e) {
+            LOG.error("Zookeeper error when getting consumer nodes", e);
+            throw new NakadiRuntimeException(e);
+        }
+
+        if (children.size() >= maxConnections) {
+            return Optional.empty();
+        }
+
+        final List<String> occupiedSlots = children;
+        final List<String> availableSlots = SLOT_NAMES.stream()
+                .filter(slot -> !occupiedSlots.contains(slot))
+                .collect(Collectors.toList());
+        final int slotIndex = new Random().nextInt(availableSlots.size());
+        final String slot = availableSlots.get(slotIndex);
```

Comment:
Very good catch. I fixed it https://github.com/zalando/nakadi/pull/489/commits/dedca7c5a89bb40f5d8b78f5407819e0f14a8aae